src/ray/raylet/node_manager.cc

```diff
@@ -1765,79 +1781,131 @@ bool NodeManager::AssignTask(const Task &task) {
+    if (client_entry == worker_clients_.end()) {
+      // TODO(atumanov): caller must handle failure to ensure tasks are not lost.
+      RAY_LOG(INFO) << "No worker found for client id " << client_id;
+      return false;
+    }
+    auto &client = client_entry->second;
+
+    rpc::AssignTaskRequest request;
+    request.set_task_id(task_id.Binary());
+    request.set_task_spec(task.Serialize());
+    request.set_resource_ids(resource_id_set.Serialize());
+
+    auto status = client->AssignTask(
+        request, [this, worker](Status status, const rpc::AssignTaskReply &reply) {
+          // Worker has finished this task. The logic below is the same
+          // as ProcessGetTaskMessage().
+          auto conn = worker->Connection();
+          auto registered_worker = worker_pool_.GetRegisteredWorker(conn);
+          if (registered_worker == nullptr) {
+            RAY_LOG(DEBUG) << "worker with pid " << worker->Pid() << " has already dead";
+            return;
           }
 
-          // Mark the task as running.
-          // (See design_docs/task_states.rst for the state transition diagram.)
-          local_queues_.QueueTasks({assigned_task}, TaskState::RUNNING);
-          // Notify the task dependency manager that we no longer need this task's
-          // object dependencies.
-          RAY_CHECK(task_dependency_manager_.UnsubscribeDependencies(spec.TaskId()));
-        } else {
-          RAY_LOG(WARNING) << "Failed to send task to worker, disconnecting client";
-          // We failed to send the task to the worker, so disconnect the worker.
-          ProcessDisconnectClientMessage(worker->Connection());
-          // Queue this task for future assignment. We need to do this since
-          // DispatchTasks() removed it from the ready queue. The task will be
-          // assigned to a worker once one becomes available.
-          // (See design_docs/task_states.rst for the state transition diagram.)
-          local_queues_.QueueTasks({assigned_task}, TaskState::READY);
-          DispatchTasks(MakeTasksWithResources({assigned_task}));
-        }
-      });
+          // If the worker was assigned a task, mark it as finished.
+          RAY_CHECK(!worker->GetAssignedTaskId().IsNil());
+          FinishAssignedTask(*worker);
+
+          HandleWorkerAvailable(worker);
+        });
+
+    // NOTE: we cannot directly call `finish_assign_task_callback` here because
+    // it assumes the task is in SWAP queue, thus we need to delay invoking this
+    // function after the assigned tasks are moved from READY queue to SWAP queue
+    // in `DispatchTasks`.
+    // Another option is to move the tasks to SWAP queue here just before calling
+    // `finish_assign_task_callback` so we can save an io_service post, at the
+    // expense of calling `MoveTask` for each of the assigned tasks.
+    io_service_.post(
+        [status, finish_assign_task_callback]() { finish_assign_task_callback(status); });
```

Comment:
Does `status == ok` ensures that the request is successfully sent to the worker? 
I guess perhaps not, because `AssignTask` is async.

Reply:
I think if server is already dead, then the client should be able to tell from the status. I can double check this. On the other side, server still can crash before it has any chance to process the request.

Reply:
Verified that the status can tell server failure if server crashes. I feel it's good enough in this case?

Reply:
I think as long as we can make sure the task won't be lost, it should be fine.

Reply:
BTW, we can add a TODO here saying that when we fully migrate to push mode, this `post` and the `swap` queue won't be needed.